### PR TITLE
Add support for Unit in similar fashion to Vaadin 8

### PR DESCRIPTION
**Pull request by TatuLund**
When we have been doing migration assessments of various customer projects, we have learned that usage of methods like setWidth(float width, Unit unit) in Vaadin 8 are rather popular. In order to make migration of such applications easier and support legacy style of setting sizes of components it is useful to have this feature in flow HasSize as well.

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java [hunk 1]
```diff
@@ -245,4 +344,129 @@ default void setSizeUndefined() {
...
+            String size = "0";
+            String unit = "";
+            try {
+                unit = units.findFirst().get().toString();
+            } catch (NoSuchElementException e) {
+                throw new IllegalArgumentException(
+                        "The parameter string '"+cssSize+"' does not contain valid unit");
+            }
+            size = cssSize.substring(0,cssSize.length()-unit.length());
+            if (size.isEmpty()) size = "0";
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/HasSize.java]: Fixed

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java [hunk 2]
```diff
@@ -245,4 +344,129 @@ default void setSizeUndefined() {
...
+         * @return A unit or null if no supported unit.
+         */
+        public static Unit getUnit(String cssSize) {
+            if (cssSize == null || cssSize.length() < 1) {
+                throw new IllegalArgumentException(
+                       "The parameter can't be null or too short to contain unit");
+            }
+            Stream<Unit> units = getUnits().filter(
+                    unit -> cssSize.endsWith(unit.toString()));
+            return units.findFirst().get();
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/HasSize.java]: Fixed

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java [hunk 3]
```diff
@@ -245,4 +344,129 @@ default void setSizeUndefined() {
...
+        public static float getSize(String cssSize) {
+            if (cssSize == null || cssSize.length() < 1) {
+                throw new IllegalArgumentException("The parameter can't be null");
+            }
+            Stream<Unit> units = getUnits().filter(
+                    unit -> cssSize.endsWith(unit.toString()));
+            String size = "0";
+            String unit = "";
+            try {
+                unit = units.findFirst().get().toString();
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/HasSize.java:441]: It is in try - catch for a purpose

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java [hunk 4]
```diff
@@ -245,4 +344,129 @@ default void setSizeUndefined() {
...
+            } catch (NoSuchElementException e) {
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/HasSize.java:442]: It is rethrown as more user friendly exception

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java [hunk 5]
```diff
@@ -245,4 +345,135 @@ default void setSizeUndefined() {
...
+            for (Unit unit : Unit.values()) {
+                if (symbol.equals(unit.getSymbol())) {
+                    return unit;
+                }
+            }
+            return Unit.PIXELS; // Defaults to pixels
+        }
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/HasSize.java:475]: Fixed

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java [hunk 6]
```diff
@@ -245,4 +344,129 @@ default void setSizeUndefined() {
         setWidth(null);
         setHeight(null);
     }
+
+    public enum Unit {
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/HasSize.java:352]: Fixed

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java [hunk 7]
```diff
@@ -245,4 +344,129 @@ default void setSizeUndefined() {
...
+         * 
+         * @return Size as float, 0 if string contained only the unit.
+         */
+        public static float getSize(String cssSize) {
+            if (cssSize == null || cssSize.length() < 1) {
+                throw new IllegalArgumentException("The parameter can't be null");
+            }
+            Stream<Unit> units = getUnits().filter(
+                    unit -> cssSize.endsWith(unit.toString()));
+            String size = "0";
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/HasSize.java]: Fixed

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java [hunk 8]
```diff
@@ -245,4 +344,129 @@ default void setSizeUndefined() {
...
+            } catch (NoSuchElementException e) {
+                throw new IllegalArgumentException(
+                        "The parameter string '"+cssSize+"' does not contain valid unit");
+            }
+            size = cssSize.substring(0,cssSize.length()-unit.length());
+            if (size.isEmpty()) size = "0";
+            return Float.valueOf(size);
+        }
+
+        public static Unit getUnitFromSymbol(String symbol) {
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/HasSize.java:462]: Fixed

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java [hunk 9]
```diff
@@ -245,4 +345,144 @@ default void setSizeUndefined() {
         setWidth(null);
         setHeight(null);
     }
+
+    /**
+     * Enum of supported units in Css sizes
+     */
+    public enum Unit {
```
Comment (caalador) [inline flow-server/src/main/java/com/vaadin/flow/component/HasSize.java:352]: Move this out from the interface to its own class as it's long and has actual code inside it.

File: flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java [hunk 10]
```diff
@@ -168,4 +225,53 @@ public void setSizeUndefined() {
         Assert.assertNull(component.getHeight());
     }
 
+    @Test
+    public void getUnit() {
+        HasSizeComponent c = new HasSizeComponent();
```
Comment (caalador) [inline flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java:230]: Remove unused variable in test.

File: flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java [hunk 11]
```diff
@@ -168,4 +225,53 @@ public void setSizeUndefined() {
...
+            String cssSize = 100f+unit.toString();
+            Optional<Unit> theUnit = Unit.getUnit(cssSize);
+            Assert.assertTrue(theUnit.isPresent());
+            Assert.assertEquals(theUnit.get(), unit);
+        }
+    }
+
+    @Test
+    public void getSize() {
+        HasSizeComponent c = new HasSizeComponent();
```
Comment (caalador) [inline flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java:241]: Remove unused variable in test.

File: flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java [hunk 12]
```diff
@@ -168,4 +225,53 @@ public void setSizeUndefined() {
...
+    public void getSizeNoValidUnit() {
+        String cssSize = "100p";
+        float size = Unit.getSize(cssSize);        	
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void getUnitException() {
+        String cssSize = null;
+        Optional<Unit> size = Unit.getUnit(cssSize);        	
+    }
```
Comment (caalador) [inline flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java:276]: Unit tests should be in their own test class

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java [hunk 13]
```diff
@@ -245,4 +345,144 @@ default void setSizeUndefined() {
         setWidth(null);
         setHeight(null);
     }
+
+    /**
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/HasSize.java:345]: Done

File: flow-server/src/main/java/com/vaadin/flow/component/HasSize.java (hunk 9 above)
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/HasSize.java]: Done

File: flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java (hunk 11 above)
Comment (TatuLund) [reply flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java]: Done

File: flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java (hunk 12 above)
Comment (TatuLund) [reply flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java]: Done

File: flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java (hunk 10 above)
Comment (TatuLund) [reply flow-server/src/test/java/com/vaadin/flow/component/HasSizeTest.java]: Done

File: flow-server/src/main/java/com/vaadin/flow/component/Unit.java [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+     * @param cssSize Css compliant size string such as "50px".
+     * 
+     * @return A Optional unit.
+     */
+    public static Optional<Unit> getUnit(String cssSize) {
+        if (cssSize == null) {
+             throw new IllegalArgumentException("The parameter can't be null");
+        }
+        Stream<Unit> units = getUnits().filter(unit -> cssSize.endsWith(unit.toString()));
+        return units.findFirst();
```
Comment (caalador) [inline flow-server/src/main/java/com/vaadin/flow/component/Unit.java:99]:
```suggestion
        return getUnits().filter(unit -> cssSize.endsWith(unit.toString())).findFirst();
```

File: flow-server/src/main/java/com/vaadin/flow/component/Unit.java [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        Stream<Unit> units = getUnits().filter(unit -> cssSize.endsWith(unit.toString()));
+        String size;
+        String unit = "";
+        try {
+            unit = units.findFirst().get().toString();
+        } catch (NoSuchElementException e) {
+            throw new IllegalArgumentException(
+                    "The parameter string '"+cssSize+"' does not contain valid unit");
+        }
+        size = cssSize.substring(0,cssSize.length()-unit.length());
```
Comment (caalador) [inline flow-server/src/main/java/com/vaadin/flow/component/Unit.java:122]:
```suggestion
        Unit unit = getUnits()
                .filter(value -> cssSize.endsWith(value.toString())).findFirst()
                .orElseThrow(() -> new IllegalArgumentException(String.format(
                        "The parameter string '%s' does not contain valid unit",
                        cssSize)));
        String size = cssSize
                .substring(0, cssSize.length() - unit.toString().length());
```

File: flow-server/src/main/java/com/vaadin/flow/component/Unit.java [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+    INCH("in"),
+    /**
+     * Unit code representing in percentage of the containing element
+     * defined by terminal.
+     */
+    PERCENTAGE("%");
+
+    private final String symbol;
+
+    private Unit(String symbol) {
```
Comment (caalador) [inline flow-server/src/main/java/com/vaadin/flow/component/Unit.java:70]:
```suggestion
    Unit(String symbol) {
```

File: flow-server/src/main/java/com/vaadin/flow/component/Unit.java (hunk 15 above)
Comment (TatuLund) [reply flow-server/src/main/java/com/vaadin/flow/component/Unit.java:122]: Ok, tried out this on my test app, seems to pass my unit tests too

**Comment by knoobie** [comment]
@pleku are there any plans to cherry pick this change to 14.5? Like Tatu said, this is often used in V8 application and would make the migration easier.

**Comment by BenWilsonI** [comment]
Great additions but we are still missing an equivalent to the getHeightUnits and getWeightUnits methods from V8's Sizeable. If I do mc.setWidth(5, Unit.Inches) and then do a Unit.getUnitFromSymbol(mc.getWidth()) it interprets "5.0in" as px. So getting the units is still finnicky

**Comment by pleku** [comment]
> @pleku are there any plans to cherry pick this change to 14.5? Like Tatu said, this is often used in V8 application and would make the migration easier.

Our apologies, this should have been marked to be picked and has been an oversight from us that it hasn't.

**Comment by pleku** [comment]
(I don't think the bot will be able to pick these so have to be manually picked)

**Pull request merged**